File: Group1/Source Code/RiskAnalysisTool.Calculation/Calculation/Utilities/DynamicVisitor.cpp
#include <pch.h>
#include "DynamicVisitor.hpp"

using namespace System;
using namespace System::Linq;
using namespace System::Collections::Generic;
using namespace System::Reflection;
using namespace RiskAnalysisTool::Calculation::Utilities;



generic <typename T>
void DynamicVisitor<T>::Visit(T target) {
	Type ^visitorType = this->GetType();

	visitorType->InvokeMember(
		"OnVisit",
		BindingFlags::InvokeMethod | BindingFlags::NonPublic | BindingFlags::Public | BindingFlags::Instance,
		nullptr,
		this,
		gcnew array < Object ^ > {target});
}

File: Group1/Source Code/RiskAnalysisTool.Common/Utilities/DynamicVisitor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;

namespace RiskAnalysisTool.Utilities
{
    public abstract class DynamicVisitor<T>
    {
        public void Visit(T target)
        {
            Type visitorType = this.GetType();
            var q = from m in visitorType.GetRuntimeMethods()
                    where m.Name == "OnVisit" && !m.IsStatic && m.ReturnType == typeof(void)
                    let parameters = m.GetParameters()
                    where 
                    parameters.Length == 1
                    let p = parameters[0]
                    where 
                    !p.IsOut && p.ParameterType.GetTypeInfo().IsAssignableFrom(target.GetType().GetTypeInfo())
                    select m;

            var method = q.FirstOrDefault();

            if (method == null)
            {
                throw new ArgumentException("Invalid target type.", "target");
            }

            try
            {
                method.Invoke(this, new object[] {target});
            }
            catch (TargetInvocationException ex)
            {
                throw ex.InnerException;
            }
        }
    }
}
